compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -2179,13 +2179,20 @@ pub fn recursive_type_with_infinite_size_error(
         err.multipart_suggestion(
             &msg,
             spans
-                .iter()
-                .flat_map(|&span| {
-                    vec![
-                        (span.shrink_to_lo(), "Box<".to_string()),
-                        (span.shrink_to_hi(), ">".to_string()),
-                    ]
-                    .into_iter()
+                .into_iter()
+                .flat_map(|(ty, span)| match ty.kind() {
+                    ty::Adt(def, args) if Some(def.did) == tcx.get_diagnostic_item(sym::Option) => {
+                        vec![(
+                            span,
+                            format!("Option<Box<{}>>", args.iter().next().unwrap().expect_ty()),
```

Comment:
The problem I see with this is that it will cause trouble with any situation where `Option` has been either shadowed or been locally renamed. Of course, we could make the same argument about `Box` in the original...